File: github_code/lru-cache/cpp/21.c
/**
 * Author:Jincheng
 * Date:180418
 * Complexity: Time O(1) Space O(N)
 * Description: List + unordered_map
 *              http://www.cnblogs.com/grandyang/p/4587511.html
 *              https://leetcode.com/problems/lru-cache/discuss/46223/O(1)-unordered_map-+-list-+-splice
 *              https://leetcode.com/problems/lru-cache/discuss/45912/Clean-Short-Standard-C++-solution-NOT-writing-C-in-C++-like-all-other-lengthy-ones
 */

class LRUCache {

private:
    
    // capacity of list
    int capacity;
    // List<pair<key,value>>
    list<pair<int,int>> items;
    // map item to it's position in list
    unordered_map<int, list<pair<int,int>>::iterator> cache;
    
public:
    LRUCache(int capacity):capacity(capacity) {
        
    }
    
    int get(int key) {
        
        auto it = cache.find(key);
        
        // the key dont't exit
        if(it == cache.end())
            return -1;
        
        // move the item to the beginning of list
        items.splice(items.begin(),items,it->second);
        return it->second->second;
    }
    
    void put(int key, int value) {
        
        auto it = cache.find(key);
        
        // if the item exist, erase it firstly
        if(it != cache.end())
        {
            items.erase(it->second);
        }
        // put the item at the beginning of the list
        items.push_front(make_pair(key,value));
        
        // update the cache
        cache[key] = items.begin();
        
        // if beyong the capacity, erase the last item
        if(items.size() > capacity)
        {
            auto key = items.rbegin()->first;
            items.pop_back();
            cache.erase(key);
        }
        
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: github_code/lru-cache/java/21.java
class LRUCache {
    
    HashMap<Integer, Integer> lru;
    LinkedList<Integer> queue;
    int capacity;
    
    public LRUCache(int capacity) {
        this.capacity = capacity;
        lru = new HashMap<>();
        queue = new LinkedList<>();
    }
    
    public int get(int key) {
        if (!lru.containsKey(key)) {
            return -1;
        }
        int index = queue.indexOf(key);
        queue.remove(index);
        queue.add(key);
        return lru.get(key);        
    }
    
    public void put(int key, int value) {
        if (lru.containsKey(key)) {
            lru.replace(key, value);
            int index = queue.indexOf(key);
            queue.remove(index);
            queue.add(key);
        }
        else {
            if (lru.size() < capacity) {
                lru.put(key, value);
                queue.add(key);
            }
            else {
                int index = queue.poll();
                lru.remove(index);
                lru.put(key, value);
                queue.add(key);
            }
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
